iceoryx_hoofs/source/posix_wrapper/semaphore_interface.cpp

```diff
@@ -0,0 +1,169 @@
+        switch (result.get_error().errnum)
+        {
+        case EINVAL:
+            LogError() << "The semaphore handle is no longer valid. This can indicate a corrupted system.";
+            return cxx::error<SemaphoreError>(SemaphoreError::INVALID_SEMAPHORE_HANDLE);
+        case EINTR:
+            LogError() << "The sem_timedwait was interrupted multiple times by the operating system. Abort operation!";
+            return cxx::error<SemaphoreError>(SemaphoreError::INTERRUPTED_BY_SIGNAL_HANDLER);
+        default:
+            LogError() << "This should never happen. An unknown error occurred.";
+            break;
+        }
+        return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+    }
+
+    return cxx::success<SemaphoreWaitState>((result.value().errnum == ETIMEDOUT) ? SemaphoreWaitState::TIMEOUT
+                                                                                 : SemaphoreWaitState::NO_TIMEOUT);
+}
+
+template <typename SemaphoreChild>
+cxx::expected<bool, SemaphoreError> SemaphoreInterface<SemaphoreChild>::tryWait() noexcept
+{
+    auto result = posixCall(iox_sem_trywait)(getHandle()).failureReturnValue(-1).ignoreErrnos(EAGAIN).evaluate();
+
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EINVAL:
+            LogError() << "The semaphore handle is no longer valid. This can indicate a corrupted system.";
+            return cxx::error<SemaphoreError>(SemaphoreError::INVALID_SEMAPHORE_HANDLE);
+        case EINTR:
+            LogError() << "The sem_trywait was interrupted multiple times by the operating system. Abort operation!";
+            return cxx::error<SemaphoreError>(SemaphoreError::INTERRUPTED_BY_SIGNAL_HANDLER);
+        default:
+            LogError() << "This should never happen. An unknown error occurred.";
+            break;
+        }
+        return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+    }
+
+    return cxx::success<bool>(result.value().errnum != EAGAIN);
+}
+
+template <typename SemaphoreChild>
+cxx::expected<SemaphoreError> SemaphoreInterface<SemaphoreChild>::wait() noexcept
+{
+    auto result = posixCall(iox_sem_wait)(getHandle()).failureReturnValue(-1).evaluate();
+
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EINVAL:
+            LogError() << "The semaphore handle is no longer valid. This can indicate a corrupted system.";
+            return cxx::error<SemaphoreError>(SemaphoreError::INVALID_SEMAPHORE_HANDLE);
+        case EINTR:
+            LogError() << "The sem_wait was interrupted multiple times by the operating system. Abort operation!";
+            return cxx::error<SemaphoreError>(SemaphoreError::INTERRUPTED_BY_SIGNAL_HANDLER);
+        default:
+            LogError() << "This should never happen. An unknown error occurred.";
+            break;
+        }
+        return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
```

Comment:
You are right. The first idea was that we hard define what errno's are defined for which posix call but since the differ from standard to standard and implementation to implementation it makes sense to have a generic translation here and safe code duplication.